renderer/GLSLProgram.cpp

```diff
@@ -0,0 +1,334 @@
+ * Resolves dynamic defines statements in GLSL source files.
+ * Note that the parsing is primitive and not context-sensitive. It will not respect multi-line comments
+ * or conditional preprocessor blocks!
+ * 
+ * Define directives should look like this:
+ * 
+ * #pragma tdm_define "DEF_NAME" // optional comment
+ * 
+ * If DEF_NAME is contained in defines, the line will be replaced by
+ * #define DEF_NAME <value>
+ * 
+ * Otherwise, it will be commented out.
+ */
+void ResolveDefines( std::string &source, const idDict &defines ) {
+	static const std::regex defineRegex( R"regex(^[ \t]*#[ \t]*pragma[ \t]+tdm_define[ \t]+"(.*)"[ \t]*(?:\/\/.*)?\r?$)regex" );
+	
+	std::smatch match;
+	while( std::regex_search( source, match, defineRegex ) ) {
+		std::string define( match[ 1 ].first, match[ 1 ].second );
+		auto defIt = defines.FindKey( define.c_str() );
+		if( defIt != nullptr ) {
+			std::string replacement = "#define " + define + " " + defIt->GetValue().c_str();
+			source.replace( match[ 0 ].first, match[ 0 ].second, replacement );
+		} else {
+			std::string replacement = "// #undef " + define;
+			source.replace( match[ 0 ].first, match[ 0 ].second, replacement );
+		}
+	}
+}
+
+GLuint GLSLProgramLoader::CompileShader( GLint shaderType, const char *sourceFile, const idDict &defines ) {
+	std::string source = ReadFile( sourceFile );
+	if( source.empty() ) {
+		return 0;
+	}
+
+	std::vector<std::string> sourceFiles { sourceFile };
+	ResolveIncludes( source, sourceFiles );
+	ResolveDefines( source, defines );
+
+	GLuint shader = qglCreateShader( shaderType );
+	GLint length = source.size();
+	const char *sourcePtr = source.c_str();
+	qglShaderSource( shader, 1, &sourcePtr, &length );
+	qglCompileShader( shader );
+
+	// check if compilation was successful
+	GLint result;
+	qglGetShaderiv( shader, GL_COMPILE_STATUS, &result );
+	if( result == GL_FALSE ) {
+		// display the shader info log, which contains compile errors
+		int length;
+		qglGetShaderiv( shader, GL_INFO_LOG_LENGTH, &length );
+		auto log = std::make_unique<char[]>( length );
+		qglGetShaderInfoLog( shader, length, &result, log.get() );
+		std::stringstream ss;
+		ss << "Compiling shader file " << sourceFile << " failed:\n" << log.get() << "\n\n";
+		// unfortunately, GLSL compilers don't reference any actual source files in their errors, but only
+		// file index numbers. So we'll display a short legend which index corresponds to which file.
+		ss << "File indexes:\n";
+		for( int i = 0; i < sourceFiles.size(); ++i ) {
+			ss << "  " << i << " - " << sourceFiles[i] << "\n";
+		}
+		common->Warning( ss.str().c_str() );
+
+		qglDeleteShader( shader );
+		return 0;
+	}
+
+	return shader;
+}
+
+globalPrograms_t globalPrograms { nullptr };
+
+void GLSL_InitPrograms() {
+	idDict stencilShadowDefines;
+	stencilShadowDefines.Set( "SHADOW_TYPE", "1" );
+	globalPrograms.stencilInteractionShader = GLSLProgram::Load( stencilShadowDefines, "glprogs/interaction.vs", "glprogs/interaction.fs" );
```

Comment:
How does include paths work?

Given that _glprogs_ is standard directory for all shaders, is it in include path?

Reply:
The path is relative to the game directory. So if a shader lies in glprogs, you'd have to include it via glprogs/shader_file. I did not implement any search path logic on top of that and think it might be overkill. But if we want to restrict shaders to the glprogs directory, we could also make that the base path for the include directive.